lindera/src/character_filter.rs

```diff
@@ -0,0 +1,321 @@
+use std::collections::HashMap;
+
+use lindera_core::character_filter::CharacterFilter;
+use regex::Regex;
+use serde::{Deserialize, Serialize};
+use unicode_normalization::UnicodeNormalization;
+
+use crate::{error::LinderaErrorKind, LinderaResult};
+
+pub const UNICODE_NORMALIZE_CHARACTER_FILTER_NAME: &str = "unicode_normalize";
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub enum UnidoceNormalizeKind {
+    #[serde(rename = "nfc")]
+    NFC,
+    #[serde(rename = "nfd")]
+    NFD,
+    #[serde(rename = "nfkc")]
+    NFKC,
+    #[serde(rename = "nfkd")]
+    NFKD,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct UnidoceNormalizeCharacterFilterConfig {
+    pub kind: UnidoceNormalizeKind,
+}
+
+impl UnidoceNormalizeCharacterFilterConfig {
+    pub fn new(kind: UnidoceNormalizeKind) -> Self {
+        Self { kind }
+    }
+
+    pub fn from_slice(data: &[u8]) -> LinderaResult<Self> {
+        serde_json::from_slice(data).map_err(|err| LinderaErrorKind::Deserialize.with_error(err))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct UnicodeNormalizeCharacterFilter {
+    config: UnidoceNormalizeCharacterFilterConfig,
+}
+
+impl UnicodeNormalizeCharacterFilter {
+    pub fn new(config: UnidoceNormalizeCharacterFilterConfig) -> Self {
+        Self { config }
+    }
+
+    pub fn from_slice(data: &[u8]) -> LinderaResult<Self> {
+        Ok(Self::new(
+            UnidoceNormalizeCharacterFilterConfig::from_slice(data)?,
+        ))
+    }
+}
+
+impl CharacterFilter for UnicodeNormalizeCharacterFilter {
+    fn apply(&self, text: &mut String) -> LinderaResult<()> {
+        *text = match self.config.kind {
+            UnidoceNormalizeKind::NFC => text.nfc().collect::<String>(),
+            UnidoceNormalizeKind::NFD => text.nfd().collect::<String>(),
+            UnidoceNormalizeKind::NFKC => text.nfkc().collect::<String>(),
+            UnidoceNormalizeKind::NFKD => text.nfkd().collect::<String>(),
+        };
+
+        Ok(())
+    }
+}
+
+pub const MAPPING_CHARACTER_FILTER_NAME: &str = "mapping";
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct MappingCharacterFilterConfig {
+    pub mapping: HashMap<char, char>,
```

Comment:
Can we have `HashMap<String, String>` for mapping, so that users can replace character sequences with other character sequences instead of replacing just one character with another one?
For example, Lucene's MappingCharfilter takes String -> String mapping.
https://lucene.apache.org/core/8_0_0/analyzers-common/org/apache/lucene/analysis/charfilter/NormalizeCharMap.Builder.html

Reply:
I see, I guess I should imitate Lucene.

Reply:
Fixed.
Please see the following commit:
https://github.com/lindera-morphology/lindera/pull/236/commits/c6141f4036cf0c8e923cc9c60e456a6735aa2f17